tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/editor/DropTargetXtextEditorListener.xtend

```diff
@@ -0,0 +1,311 @@
+import org.eclipse.swt.dnd.DropTargetAdapter
+import org.eclipse.swt.dnd.DropTargetEvent
+import org.eclipse.xtext.resource.XtextResource
+import org.eclipse.xtext.util.concurrent.IUnitOfWork
+import org.testeditor.aml.ComponentElement
+import org.testeditor.aml.InteractionType
+import org.testeditor.aml.TemplateText
+import org.testeditor.aml.TemplateVariable
+import org.testeditor.tcl.AbstractTestStep
+import org.testeditor.tcl.ComponentTestStepContext
+import org.testeditor.tcl.SpecificationStepImplementation
+import org.testeditor.tcl.StepContentElement
+import org.testeditor.tcl.TclModel
+import org.testeditor.tcl.TestCase
+import org.testeditor.tcl.TestStep
+import org.testeditor.tsl.StepContentText
+import org.testeditor.tsl.StepContentVariable
+import org.testeditor.aml.Component
+import org.eclipse.xtext.EcoreUtil2
+import org.eclipse.jface.text.source.SourceViewer
+import org.eclipse.jface.viewers.ISelection
+import org.eclipse.jface.util.LocalSelectionTransfer
+import org.eclipse.jface.viewers.TreeSelection
+import org.eclipse.jface.text.source.ISourceViewer
+import org.testeditor.tcl.impl.TclFactoryImpl
+import org.testeditor.tsl.impl.TslFactoryImpl
+import javax.inject.Inject
+import org.eclipse.swt.dnd.DND
+import org.eclipse.jface.text.TextSelection
+import org.eclipse.xtext.resource.ILocationInFileProvider
+import org.eclipse.xtend.lib.annotations.Accessors
+import org.testeditor.tsl.impl.StepContentTextImpl
+import org.testeditor.tcl.impl.SpecificationStepImplementationImpl
+import org.eclipse.xtext.ui.editor.contentassist.ContentAssistContext
+import org.eclipse.xtext.util.StringInputStream
+import java.util.Collections
+import org.eclipse.emf.ecore.resource.ResourceSet
+import org.eclipse.xtext.resource.XtextResource
+import java.net.URI
+
+class DropTargetXtextEditorListener extends DropTargetAdapter {
+
+	@Inject protected TclFactoryImpl tclFactory
+	@Inject protected TslFactoryImpl tslFactory
+
+	@Inject extension ILocationInFileProvider
+	@Inject protected ContentAssistContext.Factory contentAssistFactory
+	
+	@Accessors(PUBLIC_SETTER)
+	private DropTargetXtextEditor editor
+
+	override void dragEnter(DropTargetEvent event) {
+		if ("org.testeditor.tcl.dsl.Tcl" != editor.languageName) {
+			event.detail = DND.DROP_NONE
+		}
+		if (getDroppedObjectAs(InteractionType) == null) {
+			event.detail = DND.DROP_NONE
+		}
+	}
+
+	override void drop(DropTargetEvent event) {
+
+		val List<String> toFormat = newArrayList
+		val List<String> currentElement = newArrayList
+
+		editor.document.modify(
+			updateModel(toFormat, currentElement)
+		)
+		editor.document.modify(
+			formatRelevantRegion(toFormat)
+		)
+		editor.document.modify(
+			setCursorToNewElement(currentElement)
+		)
+
+	}
+	def updateModel(List<String> toFormat, List<String> currentElement) {
+
+		new IUnitOfWork.Void<XtextResource>() {
+
+			override process(XtextResource resource) throws Exception {
+				if (resource.contents.head instanceof TclModel) {
```

Comment:
If you assign `resource.contents.head` to a variable beforehand, there is no need for the cast below. Xtend will automatically cast it for you within the `if` clause and the code will get more readable.